main/src/ca/uwaterloo/flix/language/phase/extra/CodeQuality.scala

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright 2021 Magnus Madsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.extra
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.TypedAst.Predicate.{Body, Head}
+import ca.uwaterloo.flix.language.ast.TypedAst.{CatchRule, ChoiceRule, Constraint, Expression, MatchRule, SelectChannelRule}
+import ca.uwaterloo.flix.language.ast.{Symbol, Type, TypedAst}
+import ca.uwaterloo.flix.language.errors.CodeHint
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+
+object CodeQuality {
+
+  /**
+    * A list of operations that supports fusion or laziness when given pure function arguments.
+    */
+  val WantsPureArg: List[Symbol.DefnSym] = List(
```

Comment:
`private`

Reply:
Should we consider making this an annotation instead, so other libraries can use it?

Reply:
Indeed. This is more of an experiment. We should have a more general mechanism, but I don't know yet what it would look like. Suggestions?

Reply:
```
@thisFunctionRunsFasterIfTheFunction(f)IsPure
def map(f: a -> b)
```

🤣

Reply:
I was thinking the annotation would go on the fparam.

```scala
def map(@preferPure f: a -> b & e)

Reply:
If this is successful we can come up with some annotation.